ledger/onlineaccountscache.go

```diff
@@ -0,0 +1,94 @@
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+type onlineAccountsCache struct {
+	accounts map[basics.Address]*persistedOnlineAccountDataList
+}
+
+// init initializes the onlineAccountsCache for use.
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) init() {
+	o.accounts = make(map[basics.Address]*persistedOnlineAccountDataList)
+}
+
+// read the persistedAccountData object that the cache has for the given address.
+// thread locking semantics : read lock
+func (o *onlineAccountsCache) read(addr basics.Address, rnd basics.Round) (data persistedOnlineAccountData, has bool) {
+	if list := o.accounts[addr]; list != nil {
+		node := list.back()
+		if node.Value.round > rnd {
+			return persistedOnlineAccountData{}, false
+		}
+		for node.prev != &list.root {
+			node = node.prev
+			// only need one entry that is targetRound or older
+			if node.Value.round > rnd {
+				return *node.next.Value, true
+			}
+		}
+		return *node.Value, true
+	}
+	return persistedOnlineAccountData{}, false
+}
+
+// write a single persistedAccountData to the cache
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) writeFront(acctData persistedOnlineAccountData) {
+	if _, ok := o.accounts[acctData.addr]; !ok {
+		o.accounts[acctData.addr] = newPersistedOnlineAccountList()
+	}
+	list := o.accounts[acctData.addr]
+	if list.root.next != &list.root && acctData.round <= list.root.next.Value.round {
+		return
+	}
+	o.accounts[acctData.addr].pushFront(&acctData)
+}
+
+// write a single persistedAccountData to the cache
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) writeBack(acctData persistedOnlineAccountData) {
+	if _, ok := o.accounts[acctData.addr]; !ok {
+		o.accounts[acctData.addr] = newPersistedOnlineAccountList()
+	}
+	list := o.accounts[acctData.addr]
+	if list.root.prev != &list.root && acctData.round >= list.root.prev.Value.round {
+		return
+	}
+	o.accounts[acctData.addr].pushBack(&acctData)
+}
+
+// prune trims the onlineaccountscache by only keeping entries that would give account state
+// of rounds past targetRound
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) prune(targetRound basics.Round) {
+	for _, list := range o.accounts {
+		node := list.back()
+		for node.prev != &list.root {
+			node = node.prev
+			// only need one entry that is targetRound or older
+			if node.Value.round <= targetRound {
+				list.remove(node.next)
+			}
```

Comment:
just to be clear, this will remember the last known balance of anyone who has ever voted — right? Today that should be fine, but if we ever have a very long tail of voters we might want to e.g. evict accounts with low stake and have some max # of accounts cached or something.